### PR TITLE
Support clustered redis latency metrics

### PR DESCRIPTION
**Pull request by kandogu**

### CODE REVIEW

**Comment by sdelamo** [comment]
@kandogu can you describe the PR? What issue are you trying to solve? what feature are you adding?

File: redis-lettuce/src/main/java/io/micronaut/configuration/lettuce/DefaultRedisClusterClientFactory.java [hunk 1]
```diff
@@ -40,45 +37,30 @@
...
     @Bean(preDestroy = "shutdown")
     @Singleton
     @Primary
-    public RedisClusterClient redisClient(@Primary AbstractRedisConfiguration config, @Primary @Nullable ClientResources defaultClientResources) {
-        List<RedisURI> uris = config.getUris();
-        if (CollectionUtils.isEmpty(uris)) {
-            throw new ConfigurationException("Redis URIs must be specified");
-        }
-        return defaultClientResources == null ? RedisClusterClient.create(uris) : RedisClusterClient.create(defaultClientResources, uris);
+    public RedisClusterClient redisClient(@Primary AbstractRedisConfiguration config,
```
Comment (sdelamo) [inline redis-lettuce/src/main/java/io/micronaut/configuration/lettuce/DefaultRedisClusterClientFactory.java:46]: Deleting the old constructor would be a breaking change. You will need to deprecated it.

**Review by sdelamo — CHANGES_REQUESTED** [review]

**Comment by kandogu** [comment]
Latency metrics are not supported in the clustered client of Redis, I was trying to add latency metrics to the clustered client too as the standalone Redis client. @sdelamo

File: redis-lettuce/src/main/java/io/micronaut/configuration/lettuce/DefaultRedisClusterClientFactory.java (hunk 1 above)
Comment (kandogu) [reply redis-lettuce/src/main/java/io/micronaut/configuration/lettuce/DefaultRedisClusterClientFactory.java:46]: Actually, I didn't delete the old constructor, I implemented it in [AbstractRedisClusterClientFactory.java](https://github.com/micronaut-projects/micronaut-redis/pull/244/files/6173905cb7d7b935eec544650d86feece2cdd1c7?diff=split&w=1#diff-9e98db04d68a3d2e9810c1652bd1a4a4a0c92339b293a94b3f501a417284cc5b), (lines 41-44) as deprecated function.

Comment (kandogu) [reply redis-lettuce/src/main/java/io/micronaut/configuration/lettuce/DefaultRedisClusterClientFactory.java:46]: This class extends [AbstractRedisClusterClientFactory.java](https://github.com/micronaut-projects/micronaut-redis/pull/244/files/6173905cb7d7b935eec544650d86feece2cdd1c7?diff=split&w=1#diff-9e98db04d68a3d2e9810c1652bd1a4a4a0c92339b293a94b3f501a417284cc5b).

**Comment by kandogu** [comment]
@[sdelamo](https://github.com/sdelamo) can you look at it again?

File: redis-lettuce/src/main/java/io/micronaut/configuration/lettuce/DefaultRedisClusterClientFactory.java [hunk 2]
```diff
@@ -40,45 +37,30 @@
...
-     * @param redisClient client.
-     * @return connection
-     */
+    @Override
     @Bean(preDestroy = "close")
     @Singleton
     @Primary
     public StatefulRedisClusterConnection<String, String> redisConnection(@Primary RedisClusterClient redisClient) {
-        return redisClient.connect();
+        return super.redisConnection(redisClient);
```
Comment (guillermocalvo) [inline redis-lettuce/src/main/java/io/micronaut/configuration/lettuce/DefaultRedisClusterClientFactory.java:57]: Why move this method to the base class, when all it does is delegate to the received `redisClient`? I don't really see the point. 🤔

File: redis-lettuce/src/main/java/io/micronaut/configuration/lettuce/DefaultRedisClusterClientFactory.java [hunk 3]
```diff
@@ -40,45 +37,30 @@
...
-     *
-     * @param redisClient redisClient
-     * @return connection
-     */
+    @Override
     @Bean(preDestroy = "close")
     @Singleton
     public StatefulRedisPubSubConnection<String, String> redisPubSubConnection(@Primary RedisClusterClient redisClient) {
-        return redisClient.connectPubSub();
+        return super.redisPubSubConnection(redisClient);
```
Comment (guillermocalvo) [inline redis-lettuce/src/main/java/io/micronaut/configuration/lettuce/DefaultRedisClusterClientFactory.java:64]: Same here; am I missing something? 🤔

File: redis-lettuce/src/main/java/io/micronaut/configuration/lettuce/AbstractRedisClusterClientFactory.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    /**
+     *
+     * @param redisClient redisClient
+     * @return connection
+     */
+    public StatefulRedisPubSubConnection<String, String> redisPubSubConnection(RedisClusterClient redisClient) {
+        return redisClient.connectPubSub();
+    }
+
+    @Nullable
```
Comment (guillermocalvo) [inline redis-lettuce/src/main/java/io/micronaut/configuration/lettuce/AbstractRedisClusterClientFactory.java:82]: `configureClientResources` is annotated as `@Nullable` but it cannot possibly return `null`, can it?

File: redis-lettuce/src/main/java/io/micronaut/configuration/lettuce/AbstractRedisClusterClientFactory.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+     */
+    public RedisClusterClient redisClient(AbstractRedisConfiguration config,
+                                          @Nullable ClientResources optionalClientResources,
+                                          @Nullable List<ClientResourcesMutator> mutators) {
+        ClientResources clientResources = configureClientResources(config, optionalClientResources, mutators);
+        List<RedisURI> uris = config.getUris();
+        if (CollectionUtils.isEmpty(uris)) {
+            throw new ConfigurationException("Redis URIs must be specified");
+        }
+        return clientResources == null ? RedisClusterClient.create(uris) : RedisClusterClient.create(clientResources, uris);
```
Comment (guillermocalvo) [inline redis-lettuce/src/main/java/io/micronaut/configuration/lettuce/AbstractRedisClusterClientFactory.java:61]: Provided that `configureClientResources` is `@NonNull`:

```suggestion
        return RedisClusterClient.create(clientResources, uris);
```

**Review by guillermocalvo — CHANGES_REQUESTED** [review]
Looks good overall, except for the lack of tests.

I personally wouldn't introduce an abstract class to inherit this functionality from, because IMO it makes the implementation a bit more complicated than it needs to be. And I don't foresee many classes extending from it; I may be wrong though.

**Review by sdelamo — CHANGES_REQUESTED** [review]
We need tests.

**Review by sdelamo — APPROVED** [review]

**Pull request merged**